tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+		-- even pages by using -sPageList=even, then the output of -sOutputFile=out%d.png
+		-- would still be out1.png, out2.png, out3.png etc.
+		local cmd = ([[rungs -dSAFER -sDEVICE=pdfwrite -dNOPAUSE -dQUIET -dBATCH -dAutoRotatePages=/None -dCompatibilityLevel="%s" -sPageList="%s" -sOutputFile="%s" "%s"]]):format(
+			pdf_version,
+			table.concat(pages, ","),
+			out_pat,
+			src_pdf
+		)
+		local succ, err = os_spawn(cmd)
+
+		-- removes generated files (in case they still exist)
+		local cleanup = function()
+			for i in ipairs(pages) do
+				local fn = out_pat:format(i)
+				-- this is cleanup -> fail silently, not throwing errors
+				if kpse.out_name_ok_silent_extended(fn) then
+					if lfs.isfile(fn) then
+						return os_rm(fn)
+					end
+				end
+			end
+		end
+
+		return succ, err, cleanup, out_pat
+	end
+end
+
+-- restricted function defined here
+local check_dimensions
+do
+	-- safe the functions/libraries needed in this restricted area
+	local pdfe = pdfe
+	-- this is not inside the function but the startup code -> can error here
+	if not pdfe then error("pdfe library is not available. This script needs to be executed with texlua.") end
+
+	---check the dimensions of the pages in `src_pdf` specified in `page_dimensions`. Reports back which dimensions match (with `tolerance`) an which don't
+	---@param src_pdf string
+	---@param page_dimensions table
+	---@param tolerance number
+	---@param force boolean
+	---@return integer[] matching_pages
+	---@return integer[] failed_pages
+	---@return string pdf_version
+	---@overload fun(src_pdf:string, page_dimensions:table, tolerance:number, force:boolean): nil, string?, nil
+	check_dimensions = function(src_pdf, page_dimensions, tolerance, force)
+		local pdf
+		if kpse.in_name_ok_silent_extended(src_pdf) then
+			pdf = pdfe.open(src_pdf)
+		else
+			return nil, ("Opening %s not permitted."):format(src_pdf)
+		end
+
+		-- collect which pages succeded the dimension check
+		local succ = {}
+		-- collect which pages failed the dimension check
+		local failed = {}
+		for _, i in ipairs(page_dimensions) do
+			local p = i.page
+			local page = pdfe.getpage(pdf, p)
+			if not page then
+				-- page not found -> skip it
+				table.insert(failed, {page=i, reason="not found"})
+			else
+				local mediabox = pdfe.getbox(page, "MediaBox")
+				local w = bp2pt(mediabox[3] - mediabox[1])
+				local h = bp2pt(mediabox[4] - mediabox[2])
+				if math.abs(w - i.width) > tolerance or math.abs(h - i.height) > tolerance and not force then
+					table.insert(failed, {page=i, reason="dimension", real_width=w, real_height=h})
+				else
+					table.insert(succ, p)
+				end
+			end
+		end
+
+		local v_major, v_minor = pdfe.getversion(pdf)
+		local pdf_version = ("%d.%d"):format(v_major, v_minor)
+
+		pdfe.close(pdf)
+
+		table.sort(succ)
+		table.sort(failed, function(a,b) return a.page.page < b.page.page end)
+		return succ, failed, pdf_version
+	end
+end
```

Comment:
Hmm, does this whole function need to be inside the “unsafe” area? I think that it might be better to just make a safe wrapper for `pdfe.open`, and then do the rest in the “restricted” area.

Reply:
But it's not only, `pdfe.open` the functions
- `pdfe.getpage`
- `pdfe.getbox`
- `pdfe.getversion`
- `pdfe.close`
are also needed.

But can the `pdfe` library actually do any harm, or wouldn't it be safe to expose it to the restricted area (as a safe to use library) as well.

At least it can only do reading, no writing from what I see.

Reply:
`pdfe.open` should be safe inside a wrapper that checks `kpse.in_name_ok_silent_extended` first, and `pdfe.getpage`, `pdfe.getbox`, `pdfe.getversion`, and `pdfe.close` should be safe to expose directly.

> But can the `pdfe` library actually do any harm, or wouldn't it be safe to expose it to the restricted area (as a safe to use library) as well.

I think that it's probably safest to only copy over these 4 functions that you're actually using.

Reply:
I agree, implemented in 876c115